app/cdap/components/shared/JsonDiff/index.tsx

```diff
@@ -0,0 +1,172 @@
+  overflow: auto;
+  div {
+    width: 45%;
+    display: inline-block;
+    vertical-align: top;
+    margin-left: 1em;
+    margin-top: 0.5em;
+  }
+`;
+
+const RedSpan = styled.span`
+  display: block;
+  background-color: #feebe9;
+`;
+
+const GreenSpan = styled.span`
+  display: block;
+  background-color: #e6ffec;
+`;
+
+const NormalSpan = styled.span`
+  display: block;
+`;
+
+const EmptyLineSpan = styled.span`
+  display: block;
+  background-color: #e8e8e8;
+`;
+
+interface IJsonDiffProps {
+  isOpen: boolean;
+  onClose: () => void;
+  loading?: boolean;
+  error: string;
+  latestConfig: object;
+  selectedConfig: object;
+}
+
+const JsonDiff = ({
+  isOpen,
+  onClose,
+  loading,
+  error,
+  latestConfig,
+  selectedConfig,
+}: IJsonDiffProps) => {
+  const JsonDiffContent = () => {
+    if (!selectedConfig || !latestConfig) {
+      return;
+    }
+    const selectedConfigArray = JSON.stringify(selectedConfig, null, 2).split('\n');
+    const latestConfigArray = JSON.stringify(latestConfig, null, 2).split('\n');
+    const result = diff(selectedConfigArray, latestConfigArray);
+    if (!result) {
+      // two json are the same
+      return <h3 style={{ textAlign: 'center' }}>{T.translate(`${PREFIX}.noDiffText`)}</h3>;
+    }
+    const coloredSelectedConfig = [];
+    const coloredLatestConfig = [];
+    const selectedConfigLineNums = [];
+    const latestConfigLineNums = [];
+    let selectedConfigIdx = 0;
+    let latestConfigIdx = 0;
+    for (const array of result) {
+      if (array.length === 1) {
+        // line is same
+        coloredSelectedConfig.push(styleSpanNormal(selectedConfigArray[selectedConfigIdx]));
+        coloredLatestConfig.push(styleSpanNormal(latestConfigArray[latestConfigIdx]));
+        selectedConfigLineNums.push(styleSpanNormal(selectedConfigIdx));
+        latestConfigLineNums.push(styleSpanNormal(latestConfigIdx));
+        selectedConfigIdx++;
+        latestConfigIdx++;
+        continue;
+      }
+      if (array[0] === '+') {
+        // left should append empty line
+        coloredSelectedConfig.push(<EmptyLineSpan> </EmptyLineSpan>);
+        coloredLatestConfig.push(styleSpanStringGreen(latestConfigArray[latestConfigIdx]));
+        selectedConfigLineNums.push(styleSpanNormal(' '));
+        latestConfigLineNums.push(styleSpanNormal(latestConfigIdx));
+        latestConfigIdx++;
+        continue;
+      }
+      if (array[0] === '-') {
+        // right should append empty line
+        coloredSelectedConfig.push(styleSpanStringRed(selectedConfigArray[selectedConfigIdx]));
+        coloredLatestConfig.push(<EmptyLineSpan> </EmptyLineSpan>);
+        selectedConfigLineNums.push(styleSpanNormal(selectedConfigIdx));
+        latestConfigLineNums.push(styleSpanNormal(' '));
+        selectedConfigIdx++;
+      }
+    }
+    return (
+      <>
+        <div>
+          <b>{T.translate(`${PREFIX}.older`)}</b>
+          <pre>
+            <div style={{ float: 'left', width: 'auto' }}>
```

Comment:
These two divs just float left next to each other? I was going to ask if it breaks when a line overflows but it looks like we just cut those lines out? We don't need to show the whole text of each line?

Reply:
it horizontally overflows as well